File: memory_hub_app/lib/models/memory.dart
class Memory {
  final String id;
  final String title;
  final String content;
  final List<String> mediaUrls;
  final List<String> tags;
  final String privacy;
  final Map<String, double>? location;
  final String? mood;
  final String ownerId;
  final String? ownerName;
  final String? ownerAvatar;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int viewCount;
  final int likeCount;
  final int commentCount;
  final bool isLiked;
  final bool isBookmarked;

  Memory({
    required this.id,
    required this.title,
    required this.content,
    this.mediaUrls = const [],
    this.tags = const [],
    this.privacy = 'private',
    this.location,
    this.mood,
    required this.ownerId,
    this.ownerName,
    this.ownerAvatar,
    required this.createdAt,
    required this.updatedAt,
    this.viewCount = 0,
    this.likeCount = 0,
    this.commentCount = 0,
    this.isLiked = false,
    this.isBookmarked = false,
  });

  factory Memory.fromJson(Map<String, dynamic> json) {
    return Memory(
      id: json['id'] ?? json['_id'] ?? '',
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      mediaUrls: json['media_urls'] != null
          ? List<String>.from(json['media_urls'])
          : [],
      tags: json['tags'] != null ? List<String>.from(json['tags']) : [],
      privacy: json['privacy'] ?? 'private',
      location: json['location'] != null
          ? Map<String, double>.from(json['location'])
          : null,
      mood: json['mood'],
      ownerId: json['owner_id'] ?? '',
      ownerName: json['owner_name'],
      ownerAvatar: json['owner_avatar'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : DateTime.now(),
      viewCount: json['view_count'] ?? 0,
      likeCount: json['like_count'] ?? 0,
      commentCount: json['comment_count'] ?? 0,
      isLiked: json['is_liked'] ?? false,
      isBookmarked: json['is_bookmarked'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'content': content,
      'media_urls': mediaUrls,
      'tags': tags,
      'privacy': privacy,
      'location': location,
      'mood': mood,
      'owner_id': ownerId,
      'owner_name': ownerName,
      'owner_avatar': ownerAvatar,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'view_count': viewCount,
      'like_count': likeCount,
      'comment_count': commentCount,
      'is_liked': isLiked,
      'is_bookmarked': isBookmarked,
    };
  }
}

class MemoryCreate {
  final String title;
  final String content;
  final List<String> tags;
  final String privacy;
  final String? location;
  final String? mood;

  MemoryCreate({
    required this.title,
    required this.content,
    this.tags = const [],
    this.privacy = 'private',
    this.location,
    this.mood,
  });

  // Note: This method is not currently used. Tags are JSON-encoded in ApiService.createMemory
  Map<String, dynamic> toFormData() {
    return {
      'title': title,
      'content': content,
      'tags': tags,  // Keep as list for potential future use
      'privacy': privacy,
      if (location != null) 'location': location,
      if (mood != null) 'mood': mood,
    };
  }
}

File: app/models/memory.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from enum import Enum
from pydantic import BaseModel, Field, HttpUrl, validator, field_validator
from bson import ObjectId
from .user import PyObjectId

class MemoryPrivacy(str, Enum):
    PRIVATE = "private"
    FRIENDS = "friends"
    PUBLIC = "public"

class MemoryBase(BaseModel):
    title: str
    content: str
    media_urls: List[str] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)
    privacy: MemoryPrivacy = MemoryPrivacy.PRIVATE
    location: Optional[Dict[str, float]] = None  # { "lat": 0.0, "lng": 0.0 }
    mood: Optional[str] = None
    weather: Optional[Dict[str, Any]] = None
    tagged_family_members: List[Dict[str, str]] = Field(default_factory=list)  # [{"user_id": "xxx", "relation": "mom"}]
    family_circle_ids: List[str] = Field(default_factory=list)  # Family circles this memory is shared with
    relationship_context: Optional[str] = None  # e.g., "Mom's Birthday", "Family Reunion"
    genealogy_person_ids: List[str] = Field(default_factory=list)  # Genealogy persons tagged in this memory
    family_tree_id: Optional[str] = None  # Family tree this memory belongs to
    
    @validator('title')
    def title_must_not_be_empty(cls, v):
        if not v.strip():
            raise ValueError('Title cannot be empty')
        return v.strip()
    
    @validator('content')
    def content_must_not_be_empty(cls, v):
        if not v.strip():
            raise ValueError('Content cannot be empty')
        return v.strip()

class MemoryCreate(MemoryBase):
    pass

class MemoryUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    media_urls: Optional[List[str]] = None
    tags: Optional[List[str]] = None
    privacy: Optional[MemoryPrivacy] = None
    location: Optional[Dict[str, float]] = None
    mood: Optional[str] = None
    genealogy_person_ids: Optional[List[str]] = None
    family_tree_id: Optional[str] = None
    
    @field_validator('title', 'content', mode='before')
    def empty_str_to_none(cls, v):
        if v == "":
            return None
        return v

class MemoryInDB(MemoryBase):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    owner_id: PyObjectId
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    view_count: int = 0
    like_count: int = 0
    comment_count: int = 0
    
    class Config:
        json_encoders = {ObjectId: str}
        populate_by_name = True
        arbitrary_types_allowed = True

class MemoryResponse(MemoryInDB):
    is_liked: bool = False
    is_bookmarked: bool = False
    owner_name: Optional[str] = None
    owner_avatar: Optional[str] = None

class MemorySearchParams(BaseModel):
    query: Optional[str] = None
    tags: Optional[List[str]] = None
    privacy: Optional[MemoryPrivacy] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    sort_by: str = "created_at"
    sort_order: str = "desc"
    page: int = 1
    limit: int = 20